examples/vmi-gpu.yaml

```diff
@@ -0,0 +1,36 @@
+---
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachineInstance
+metadata:
+  labels:
+    special: vmi-gpu
+  name: vmi-gpu
+spec:
+  domain:
+    devices:
+      disks:
+      - disk:
+          bus: virtio
+        name: containerdisk
+      - disk:
+          bus: virtio
+        name: cloudinitdisk
+      gpus:
+      - name: nvidia.com/GP102GL_Tesla_P40
```

Comment:
Ah sorry, but I have to ask you for another change. I didn't realize before that the `name` would have two purposes here: I dentifying the device on the VM (virtual representation) and also identifying the device from the device plugin perspective.

I think we can't go down that route. We can't identify the device then if multiple gpus are added and the restrictions regarding to valid names may differ. There are a few alternatives. An easy one would be adding an extra attribute like `device` or `deviceName`:

```yaml
gpus:
- name: mygpu1
  deviceName: nvidia.com/GP102GL_Tesla_P40
- name: mygpu2
  deviceName: nvidia.com/GP102GL_Tesla_P40
```

This example also illustrates the advantage of having `name` and `deviceName` separated. This way you can have two gpus from the same type attached and can always differentiate them.

Another alternative (maybe too much, if there will never be sub-configurations on the device plugin possible), would be to split it like volumes or networks into a system- and a vmi-section.

Reply:
Also note the following things: As soon as we want to support gpus in a non-devicePlugin way we will have issues evolving the api arount the `gpus` field if it is structured like I just suggested above. We don't have and don't plan to introduce `type` fields.

The general way to solve including different implementations on how to access a cpu like it is e.g. done for volumes (like here: https://github.com/kubevirt/kubevirt/blob/master/staging/src/kubevirt.io/client-go/api/v1/schema.go#L484).

So if we ever plan to have completely different `gpus` mechanisms, we should probably structure it either like this:

```yaml
spec:
  domain:
    gpus:
    - name: mygpu1
      qemutune1: a
      qemutune2: b
      devicePlugin:
        name: nvidia.com/GP102GL_Tesla_P40
        passtrhoughoption1: a
        passtrhoughoption2: a
```

or completely split it like for volumes or networks:

```yaml
spec:
  domain:
    gpus:
    - name: mygpu1
      qemutune1: a
      qemutune2: b
  [...]
  gpus:
    - name: mygpu1
      devicePlugin:
        name: nvidia.com/GP102GL_Tesla_P40
        passtrhoughoption1: a
        passtrhoughoption2: a
```

Reply:
So @visheshtanksale, @davidvossel, @vladikr, @booxter, @SchSeba, @cynepco3hahue that is  something we should keep in mind and we should be pretty sure that we will only have device plugins as a way to asign gpus.

Still the PR is pretty close. What do you think?

Reply:
The main thing I'd be concerned about here is the ability to pass in multiple gpus with the same device name.

I do agree that it would be wise to look ahead and plan for how we'd want to support growing this API in the future. At the same time, I'd hate for this to become overly complex though. I think the best thing to do is look at previous use-cases and determine what level of flexibility we really need. If we've never encountered the need to have per-gpu qemu tunings, or the ability to pass in the multiple GPUs using different methods for the same VM, then that level of flexibility might be overkill for us.

Reply:
I wonder if we could get away with just adding a `count` for a certain gpu name
```
gpus:
      - name: nvidia.com/GP102GL_Tesla_P40
        count: 2
```

I don't think we would need additional tuning for gpus. Afaik, normally, if it's a vgpu, we pre-create a device with certain parameters a give a unique name.

Reply:
> I wonder if we could get away with just adding a count for a certain gpu name

seems like that should work.

would it make sense to change `name` to something like `resourceName` or `deviceName` so it's maybe a little more explicit that value maps directly to the gpu resource that's being requested?

Reply:
I would want to separate two things here:

 * First we were waiting for a long time to have the chance to set the device `alias` upfront, and not have to rediscover it out of what qemu assigns to devices. From my point of view, having the `name` as a requirement, and having it per device is a must in our API for any device. Otherwise something like hotplug and device tracking will become very complex (like it is in e.g. ovirt). So `name` upfront and one  entry per device is from my point of view just something we need to have.
 * Second I agree that we should not make it too difficult. However because of 1 we have to look a little bit into the future.

Reply:
Seems like there needs to be discussion on API definition Kubevirt wants to have. May be we can address that as separate PR?

Reply:
We can agree to not freeze the api in the gpu section yet. Then yes. Otherwise I am not so happy, because `name` is exactly the field which we need for the device alias in the vmi side.

Reply:
And my initial plan was to let further improvements happen in a follow-up, but this is really just because of the `name` field.

Reply:
@rmohr I feel having alias is very helpful when implementing sidecars or even for debugging. So we should address `name` field in this PR. But making the gpu section of the API more rich by adding fields like `qemutune` can be added as a follow-up.

Reply:
> But making the gpu section of the API more rich by adding fields like `qemutune` can be added as a follow-up.

Agreed, that is not what I wanted to get at. My main concern is the main structure, which we should be able to evolve without deprecating fields. The device plugin is vendor neutral, so that is good, however there could be completely other ways on how to expose a gpu to a VMI (e.g. something via sidecars, ...).

If we have this situation:

```yaml
gpus:
- name: mygpu1
  deviceName: nvidia.com/GP102GL_Tesla_P40
```

we can not easily express other ways of exposing a gpu to a VMI without complicating the API. The options we would have, would be to 

1) deprecate `deviceName`, make it optional and restructure the api to this:

```yaml
gpus:
- name: mygpu1
  devicePlugin: # a substruct, like on volumes
    deviceName: nvidia.com/GP102GL_Tesla_P40
```

with this deprecated  fallback to keep the api backward compatible:

```yaml
gpus:
- name: mygpu1
  deviceName: nvidia.com/GP102GL_Tesla_P40 # deprecated but still accepted
```

other solutions can then be expressed like this:

```yaml
gpus:
- name: mygpu1
  flexGPU:
    arbitrary: attribute
```

2) add a type field, see device plugins as default and expect for all other cases the type to be set:

```yaml
gpus:
- name: mygpu1
  type: DevicePlugin # optional
  deviceName: nvidia.com/GP102GL_Tesla_P40 # this would only be required for some types
```

other solutions could then do something like this:

```yaml
gpus:
- name: mygpu1
  type: flexGPU
  arbitrary: attribute
```

Therefore I would tend to this structure from the beginning:

```yaml
gpus:
- name: mygpu1
  devicePlugin: # a substruct, like on volumes
    deviceName: nvidia.com/GP102GL_Tesla_P40 # could be `name` or `deviceName` now
```

It is more complicated than having no substruct but not as complicated as having two independent sections like for volumes and networks, but still allows a clean evolution.

Thoughts?